### ISSUE TITLE
How do you register file watchers?

### ISSUE THREAD

**Issue by Timmmm** [ebkalderon/tower-lsp#190]
The documentation says to use `client.register_capability()` but it isn't really clear how. There's also a `DidChangeWatchedFilesRegistrationOptions` struct in there but it isn't referenced from anywhere else. An example would be great!

**Comment by ebkalderon** [comment]
Good point, @Timmmm! The struct in question referenced in the documentation refers to [`lsp_types::DidChangeWatchedFilesRegistrationOptions`](https://docs.rs/lsp-types/0.74.2/lsp_types/struct.DidChangeWatchedFilesRegistrationOptions.html), and if you look at the [official LSP specification](https://microsoft.github.io/language-server-protocol/specification) (Ctrl+F for `DidChangeWatchedFilesRegistrationOptions`) you should be able to understand how it is intended to be used.

With that said, though, I totally agree we should add another example program which demonstrates the use of a simplistic file watcher with a basic server. Would you like to collaborate or open up a pull request to add such an example?

**Comment by Timmmm** [comment]
I would but it's still not clear what I actually *do* with `DidChangeWatchedFilesRegistrationOptions` - as I said the struct is defined in `lsp_types`, but it isn't actually taken as a parameter by anything. The official LSP spec doesn't really clear that up either, unless I'm missing something.

**Comment by ebkalderon** [comment]
To my understanding, you would need to use `serde_json::to_value` to convert the `DidChangeWatchedFilesRegistrationOptions` into a `serde_json::Value` and then use that value when filling out an [`lsp_types::Registration`](https://docs.rs/lsp-types/0.74.2/lsp_types/struct.Registration.html), like so:

```rust
let value = serde_json::to_value(DidChangeWatchedFilesRegistrationOptions { ... })
    .unwrap();

let registration = Registration {
    id: "myserver-fileWatcher".to_string(),
    method: "workspace/didChangeWatchedFiles".to_string(),
    register_options: Some(value),
};

client.register_capabilities(vec![registration]).await.unwrap();
```

**Comment by Timmmm** [comment]
Ah I see! Ok that isn't obvious. :-) Perhaps it would be worth adding proper functions to do this, like

```
impl Client {
  ...
  async fn register_did_change_watched_files(&self, options: &DidChangeWatchedFilesRegistrationOptions, id: &str) -> ? {
    let registration = Registration {
      id: id.to_string(),
      method: "workspace/didChangeWatchedFiles".to_string(),
      register_options: Some(options),
    };

    client.register_capabilities(vec![registration]).await
  }
```

That would be better than documentation.

**Comment by ebkalderon** [comment]
@Timmmm I'm working on this! I've submitted some pull requests to `lsp-types` adding some missing registration capabilities which should enable this feature to be added eventually. These additions are available in crates.io as of `lsp-types` 0.78. Hopefully this should be enough for work on this feature to begin.

I'm not very keen on adding dozens or possibly hundreds of specialized methods to `Client`, especially when the registration schema is likely to change frequently in between releases of the LSP spec. I think it would be better to leverage the highly generic [`lsp_types::request::Request`](https://docs.rs/lsp-types/0.79.0/lsp_types/request/trait.Request.html) and [`lsp_types::Notification`](https://docs.rs/lsp-types/0.79.0/lsp_types/notification/trait.Notification.html) traits for this instead, to avoid widespread duplication.

The design I've been contemplating is to expose the following trait in `tower-lsp`:

```rust
trait RegistrationExt {
    type Params: Serialize;

    fn build_registration<T: Into<String>>(id: T, params: Self::Params) -> Registration;
}

impl RegistrationExt for lsp_types::notification::DidChangeWatchedFiles { ... }
impl RegistrationExt for ... { ... }
```

The usage would look something like this:

```rust
use tower_lsp::RegistrationExt;
use tower_lsp::lsp_types::{notification::DidChangeWatchedFiles, *};

let value = DidChangeWatchedFilesRegistrationOptions { ... };
let registration = DidChangeWatchedFiles::build_registration("myserver-fileWatcher", value);
client.register_capabilities(vec![registration]).await.unwrap();
```

I think this is quite flexible and understandable enough without being a massive maintenance burden. Additionally, by visiting the docs page for `tower_lsp::RegistrationExt`, it would be possible for the user to see what requests/notifications support client registration and what their corresponding `RegistrationOptions` are. 😄

**Comment by ebkalderon** [comment]
Does this API look reasonable to everyone? Would this suffice for your purposes, @Timmmm? Feedback from any interested party is most welcome.

**Comment by Timmmm** [comment]
Looks like a decent API to me! Although does it prevent you accidentally doing this?

```
let value = SomeOtherRegistrationOptions { ... };
let registration = DidChangeWatchedFiles::build_registration("myserver-fileWatcher", value);
```

Maybe it is simpler just to do something like this:

```
trait Registerable: Serialize {
    fn build_registration<T: Into<String>>(&self, id: T) -> Registration;
}

impl Registerable for DidChangeWatchedFilesRegistrationOptions { ... }

let value = DidChangeWatchedFilesRegistrationOptions { ... };
let registration = value.build_registration("myserver-fileWatcher");
client.register_capabilities(vec![registration]).await.unwrap();
```

**Comment by ebkalderon** [comment]
Not actively working on it anymore, but still mulling over the proposed API in the back of my head. For now, this addition is not a particularly high priority, unless someone chimes in.

**Comment by ratmice** [comment]
Just wanted to ditto that it doesn't seem like a high priority, while perhaps not the most elegant, file watchers are possible to get working with the current API.  I'd be happy to make a PR adding an example for the current mechanism if that would be helpful.

**Comment by Feel-ix-343** [comment]
Need this!!!
